engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -255,99 +448,153 @@ private Void handleCreateVolumeFromSnapshotBothOnStorageSystem(SnapshotInfo snap
 
             VolumeApiResult result = future.get();
 
+            if (volumeDetail != null) {
+                _volumeDetailsDao.remove(volumeDetail.getId());
+            }
+
             if (result.isFailed()) {
                 s_logger.debug("Failed to create a volume: " + result.getResult());
 
                 throw new CloudRuntimeException(result.getResult());
             }
-        }
-        catch (Exception ex) {
-            throw new CloudRuntimeException(ex.getMessage());
-        }
 
-        volumeInfo = _volumeDataFactory.getVolume(volumeInfo.getId(), volumeInfo.getDataStore());
+            volumeInfo = _volumeDataFactory.getVolume(volumeInfo.getId(), volumeInfo.getDataStore());
 
-        volumeInfo.processEvent(Event.MigrationRequested);
+            volumeInfo.processEvent(Event.MigrationRequested);
 
-        volumeInfo = _volumeDataFactory.getVolume(volumeInfo.getId(), volumeInfo.getDataStore());
+            volumeInfo = _volumeDataFactory.getVolume(volumeInfo.getId(), volumeInfo.getDataStore());
 
-        HostVO hostVO = getHost(snapshotInfo.getDataStore().getId());
+            if (useCloning) {
+                copyCmdAnswer = performResignature(volumeInfo, hostVO);
+            }
+            else {
+                // asking for a XenServer host here so we don't always prefer to use XenServer hosts that support resigning
+                // even when we don't need those hosts to do this kind of copy work
+                hostVO = getHost(snapshotInfo.getDataCenterId(), false);
 
-        String value = _configDao.getValue(Config.PrimaryStorageDownloadWait.toString());
-        int primaryStorageDownloadWait = NumbersUtil.parseInt(value, Integer.parseInt(Config.PrimaryStorageDownloadWait.getDefaultValue()));
-        CopyCommand copyCommand = new CopyCommand(snapshotInfo.getTO(), volumeInfo.getTO(), primaryStorageDownloadWait, VirtualMachineManager.ExecuteInSequence.value());
+                copyCmdAnswer = performCopyOfVdi(volumeInfo, snapshotInfo, hostVO);
+            }
 
-        CopyCmdAnswer copyCmdAnswer = null;
+            if (copyCmdAnswer == null || !copyCmdAnswer.getResult()) {
+                if (copyCmdAnswer != null && !StringUtils.isEmpty(copyCmdAnswer.getDetails())) {
+                    errMsg = copyCmdAnswer.getDetails();
+                }
+                else {
+                    errMsg = "Unable to perform host-side operation";
```

Comment:
Please include more detail about the operation being performed to assist operational debugging.